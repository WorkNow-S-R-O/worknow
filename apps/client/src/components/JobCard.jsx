import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { useIntlayer } from 'react-intlayer';
import { useState } from 'react';
import useFetchCities from '../hooks/useFetchCities';
import useFetchCategories from '../hooks/useFetchCategories';
import Skeleton from 'react-loading-skeleton';
import 'react-loading-skeleton/dist/skeleton.css';
import { ImageModal } from './ui';

const JobCard = ({ job }) => {
	const content = useIntlayer('jobCard');
	const navigate = useNavigate();
	const { cities, loading: citiesLoading } = useFetchCities();
	const { categories, loading: categoriesLoading } = useFetchCategories();
	const [showImageModal, setShowImageModal] = useState(false);
	const [imageLoading, setImageLoading] = useState(true);

	// Получаем название города на нужном языке
	let cityLabel = null;
	if (job.cityId && Array.isArray(cities)) {
		const city = cities.find(
			(c) => c.value === job.cityId || c.id === job.cityId,
		);
		cityLabel = city?.label || city?.name || null;
	}

	// Получаем название категории на нужном языке
	let categoryLabel = null;
	if (job.categoryId && Array.isArray(categories)) {
		const category = categories.find(
			(c) => c.value === job.categoryId || c.id === job.categoryId,
		);
		categoryLabel = category?.label || category?.name || null;
	}

	const handleImageClick = (e) => {
		e.stopPropagation(); // Prevent card click when clicking image
		setShowImageModal(true);
	};

	const handleCloseModal = () => {
		setShowImageModal(false);
	};

	const handleImageLoad = () => {
		setImageLoading(false);
	};

	const handleImageError = () => {
		setImageLoading(false);
	};

	return (
		<>
			<div
				className={`card shadow-sm mb-4 position-relative text-start ${
					job.user?.isPremium ? 'premium-job' : ''
				}`}
				style={{
					backgroundColor: job.user?.isPremium ? '#D4E6F9' : 'white',
					width: '90%',
					maxWidth: '700px',
					borderRadius: '6px',
					minHeight: '220px',
					cursor: job.user?.clerkUserId ? 'pointer' : 'default',
				}}
				onClick={() => {
					if (job.user?.clerkUserId) {
						navigate(`/profile/${job.user.clerkUserId}`);
					}
				}}
			>
				{/* Плашка Премиум */}
				{job.user?.isPremium && (
					<div className="premium-badge">
						<i className="bi bi-star-fill"></i> {content.premiumBadge.value}
					</div>
				)}
				<div className="card-body">
					<h5 className="card-title text-primary">{job.title}</h5>
					{(job.categoryId || job.category?.label) && (
						<div className="mb-2">
							{categoriesLoading ? (
								<Skeleton
									width={80}
									height={24}
									style={{ display: 'inline-block', borderRadius: '4px' }}
								/>
							) : (
								<span className="px-2 py-1 text-sm rounded font-semibold bg-primary text-white">
									{categoryLabel ||
										job.category?.label ||
										content.notSpecified.value}
								</span>
							)}
						</div>
					)}
					<p className="card-text">
						<strong>{content.salaryPerHourCard.value}</strong>{' '}
						{job.salary || content.notSpecified.value}
						<br />
						<strong>{content.locationCard.value}</strong>{' '}
						{citiesLoading ? (
							<Skeleton
								width={80}
								height={18}
								style={{ display: 'inline-block', verticalAlign: 'middle' }}
							/>
						) : (
							cityLabel || content.notSpecified.value
						)}
					</p>
					<p className="card-text">
						{job.description || content.descriptionMissing.value}
					</p>
					<div className="card-text">
						{typeof job.shuttle === 'boolean' && (
							<div className="card-text">
								<strong>{content.shuttle.value}:</strong>{' '}
								{job.shuttle ? content.yes.value : content.no.value}
							</div>
						)}
						{typeof job.meals === 'boolean' && (
							<div className="card-text">
								<strong>{content.meals.value}:</strong>{' '}
								{job.meals ? content.yes.value : content.no.value}
							</div>
						)}
						<strong>{content.phoneNumberCard.value}</strong>{' '}
						{job.phone || content.phoneNotSpecified.value}
					</div>

					{/* Image displayed under phone number in mini size */}
					{job.imageUrl && (
						<div className="mt-3">
							{imageLoading && (
								<Skeleton
									width={120}
									height={80}
									style={{
										borderRadius: '3px',
										border: '1px solid #e0e0e0',
									}}
								/>
							)}
							<img
								src={job.imageUrl}
								alt={job.title}
								className="img-fluid rounded"
								style={{
									width: '120px',
									height: '80px',
									objectFit: 'cover',
									borderRadius: '3px',
									border: '1px solid #e0e0e0',
									cursor: 'pointer',
									display: imageLoading ? 'none' : 'block',
								}}
								onClick={handleImageClick}
								onLoad={handleImageLoad}
								onError={handleImageError}
							/>
						</div>
					)}
				</div>
			</div>

			{/* Image Modal - Only render if there's an image URL */}
			{job.imageUrl && (
				<ImageModal
					show={showImageModal}
					onHide={handleCloseModal}
					imageUrl={job.imageUrl}
					imageAlt={job.title}
					onImageError={() => {}}
				/>
			)}
		</>
	);
};

// **Валидация пропсов**
JobCard.propTypes = {
	job: PropTypes.shape({
		id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
		cityId: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
		city: PropTypes.object,
		title: PropTypes.string,
		salary: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
		description: PropTypes.string,
		phone: PropTypes.string,
		category: PropTypes.object,
		user: PropTypes.object,
		shuttle: PropTypes.bool,
		meals: PropTypes.bool,
		imageUrl: PropTypes.string,
	}).isRequired,
};

export default JobCard;
