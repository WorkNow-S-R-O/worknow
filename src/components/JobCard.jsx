import PropTypes from "prop-types";
import { useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";

const JobCard = ({ job }) => {
  const { t } = useTranslation();
  const navigate = useNavigate();

  return (
    <div
      className={`card shadow-sm mb-4 position-relative ${
        job.user?.isPremium ? "border border-warning premium-glow" : ""
      }`}
      style={{
        backgroundColor: "white",
        width: "90%",
        maxWidth: "700px",
        borderRadius: "10px",
        minHeight: "220px",
        boxShadow: job.user?.isPremium
          ? "0px 0px 15px 5px rgba(255, 215, 0, 0.7)"
          : "none",
        cursor: job.user?.clerkUserId ? "pointer" : "default",
      }}
      onClick={() => {
        if (job.user?.clerkUserId) {
          navigate(`/profile/${job.user.clerkUserId}`);
        }
      }}
    >
      <div className="card-body">
        <h5 className="card-title text-primary">{job.title}</h5>
        <p className="card-text">
          <strong>{t("salary_per_hour_card")}</strong>{" "}
          {job.salary || "Не указано"}
          <br />
          <strong>{t("location_card")}</strong> {job.city?.name || "Не указано"}
        </p>
        <p className="card-text">{job.description || "Описание отсутствует"}</p>
        <p className="card-text">
          <strong>{t("phone_number_card")}</strong> {job.phone || "Не указан"}
        </p>
      </div>
    </div>
  );
};

// **Валидация пропсов**
JobCard.propTypes = {
  job: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    salary: PropTypes.string,
    city: PropTypes.shape({
      name: PropTypes.string,
    }),
    description: PropTypes.string,
    phone: PropTypes.string,
    user: PropTypes.shape({
      clerkUserId: PropTypes.string,
      imageUrl: PropTypes.string,
      isPremium: PropTypes.bool,
      name: PropTypes.string,
    }),
  }).isRequired,
};

export default JobCard;
