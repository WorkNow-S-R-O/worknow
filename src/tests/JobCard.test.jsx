import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import JobCard from '../components/JobCard';

// Мокаем useTranslation
jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));
// Мокаем только useNavigate
jest.mock('react-router-dom', () => ({
  useNavigate: () => jest.fn(),
}));
// Мокаем Spinner из react-bootstrap
jest.mock('react-bootstrap/Spinner', () => () => <div>Spinner</div>);

const mockJob = {
  id: 1,
  title: 'Test Job',
  salary: '1000',
  city: { name: 'Москва' },
  description: 'Описание работы',
  phone: '+79999999999',
  user: {
    clerkUserId: '123',
    imageUrl: '/images/default-avatar.png',
    isPremium: false,
    name: 'Иван Иванов',
  },
};

test('renders JobCard without crashing', () => {
  const { getByText } = render(
    <JobCard job={mockJob} />
  );
  expect(getByText('Test Job')).toBeInTheDocument();
  expect(getByText('Описание работы')).toBeInTheDocument();
});

test('JobCard matches snapshot', () => {
  const { asFragment } = render(
    <JobCard job={mockJob} />
  );
  expect(asFragment()).toMatchSnapshot();
}); 