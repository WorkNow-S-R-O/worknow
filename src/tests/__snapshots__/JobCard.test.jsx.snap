// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="card shadow-sm mb-4 position-relative "
    style="background-color: white; width: 90%; max-width: 700px; border-radius: 10px; min-height: 220px; box-shadow: none;"
  >
    <div
      class="card-body"
    >
      <h5
        class="card-title text-primary"
      >
        Test Job
      </h5>
      <p
        class="card-text"
      >
        <strong>
          salary_per_hour_card
        </strong>
         1000
        <br />
        <strong>
          location_card
        </strong>
         Москва
      </p>
      <p
        class="card-text"
      >
        Описание работы
      </p>
      <p
        class="card-text"
      >
        <strong>
          phone_number_card
        </strong>
         +79999999999
      </p>
    </div>
    <div
      class="position-absolute top-0 end-0 m-2 d-flex justify-content-center align-items-center"
      style="width: 45px; height: 45px; cursor: pointer; border-radius: 50%; overflow: hidden; background-color: rgb(240, 240, 240);"
    >
      <div>
        Spinner
      </div>
      <img
        alt="User Avatar"
        class="rounded-circle w-100 h-100"
        src="/images/default-avatar.png?t=1748365211530"
        style="display: none;"
      />
    </div>
  </div>
</DocumentFragment>
`;
